Framework/API/src/Sample.cpp

```diff
@@ -363,7 +363,10 @@ int Sample::loadNexus(::NeXus::File *file, const std::string &group) {
     }
     Kernel::Material material;
     material.loadNexus(file, "material");
-    m_shape->setMaterial(material);
+    // CSGObject expected, if so, set its material
+    if (auto csgObj = boost::dynamic_pointer_cast<Geometry::CSGObject>(m_shape)) {
```

Comment:
I was speaking to @martyngigg about this last week, and I think that the main issue is that `ShapeFactory`, should not have been modified to return a `IObject`. `ShapeFactory` should create `CSGObject` only. If you make that change then we can remove these `dynamic_pointer_casts`. That should be very easy to do, and make the code cleaner.

Reply:
This may take a long time, because of the many users of `shapeFactory'.

Reply:
Do we need to update every usage?

Reply:
Doing
```
IObject *shape = shape_factory.create(...)
```
will still work as `CSGObject*` can be converted automatically to an `IObject*`. I think it's just the places where we can avoid a dynamic cast?

Reply:
Thanks for the information. That explains why nearly everything builds OK. Only `DetectorEfficiencyCorTest.h` fails.

Reply:
I also find that similar changes are needed in `ComponentCreationHelper`. I'll make those where necessary to remove the dynamic casts and possibly throughout.

Reply:
I find I've already made these changes (only 2 functions needed modifying).

Reply:
Once you've made the minor change to the `ShapeFactory` `create` to return the `CSGObject`, you should only need to remove the `dynamic_ptr_casts` you added earlier.

Reply:
This dynamic_cast can be removed, you don't need it.